src/bio/bio_dev_monitor.c

```diff
@@ -0,0 +1,442 @@
+		D_PRINT("Command ID:          0x%x\n", error_entry->cid);
+		D_PRINT("Phase bit:           %x\n", error_entry->status.p);
+		D_PRINT("Status code:         0x%x\n", error_entry->status.sc);
+		D_PRINT("Status code type:    0x%x\n", error_entry->status.sct);
+		D_PRINT("Do not retry:        %x\n", error_entry->status.dnr);
+		D_PRINT("Error location:      0x%x\n",
+			error_entry->error_location);
+		D_PRINT("LBA:                 0x%"PRIx64"\n", error_entry->lba);
+		D_PRINT("Namespace:           0x%x\n", error_entry->nsid);
+		D_PRINT("Vendor log page:     0x%x\n",
+			error_entry->vendor_specific);
+		D_PRINT("\n");
+	}
+}
+
+static void
+get_spdk_identify_ctrlr_completion(struct spdk_bdev_io *bdev_io, bool success,
+				   void *cb_arg)
+{
+	struct bio_health_monitoring			*dev_health = cb_arg;
+	struct spdk_nvme_ctrlr_data			*cdata;
+	struct spdk_io_channel				*channel;
+	struct spdk_bdev				*bdev;
+	struct spdk_bdev_desc				*health_desc;
+	struct spdk_nvme_cmd				 cmd;
+	uint32_t					 error_page_sz;
+	uint32_t					 error_page_buf_sz;
+	uint32_t					 numd, numdl, numdu;
+	int						 rc;
+	int						 sc, sct;
+
+	/* Additional NVMe status information */
+	spdk_bdev_io_get_nvme_status(bdev_io, &sct, &sc);
+	if (sc)
+		D_ERROR("NVMe status code/type: %d/%d\n", sc, sct);
+
+	/* Free I/O request in the competion callback */
+	spdk_bdev_free_io(bdev_io);
+
+	health_desc = dev_health->bhm_desc;
+	channel = spdk_bdev_get_io_channel(health_desc);
+	if (!channel) {
+		D_ERROR("Failed to get bdev I/O channel for health desc\n");
+		return;
+	}
+
+	bdev = spdk_bdev_desc_get_bdev(health_desc);
+	cdata = dev_health->bhm_ctrlr_buf;
+	/* set the number of error log page entries supported by device */
+	dev_health->bhm_elpe = cdata->elpe;
+
+
+	/*TODO Add additional relevant controller data */
+	D_PRINT("==========================================================\n");
+	D_PRINT("SPDK Device Controller Data [%s]:\n",
+		bdev != NULL ? spdk_bdev_get_name(bdev) : "");
+	D_PRINT("==========================================================\n");
+	D_PRINT("Vendor ID: %04x\n", cdata->vid);
+	D_PRINT("Serial Number: ");
+	dprint_ascii_string(cdata->sn, sizeof(cdata->sn));
+	D_PRINT("\n");
+	D_PRINT("Model Number: ");
+	dprint_ascii_string(cdata->mn, sizeof(cdata->mn));
+	D_PRINT("\n");
+	D_PRINT("Firmware Version: ");
+	dprint_ascii_string(cdata->fr, sizeof(cdata->fr));
+	D_PRINT("\n");
+	D_PRINT("Error log page entries supported: %d\n", cdata->elpe + 1);
+
+	/* Prep NVMe command to get device error log pages */
+	error_page_sz = sizeof(struct spdk_nvme_error_information_entry);
+	numd = error_page_sz / sizeof(uint32_t) - 1u;
+	numdl = numd & 0xFFFFu;
+	numdu = (numd >> 16) & 0xFFFFu;
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.opc = SPDK_NVME_OPC_GET_LOG_PAGE;
+	cmd.nsid = SPDK_NVME_GLOBAL_NS_TAG;
+	cmd.cdw10 = numdl << 16;
+	cmd.cdw10 |= SPDK_NVME_LOG_ERROR;
+	cmd.cdw11 = numdu;
+	error_page_buf_sz = error_page_sz * (dev_health->bhm_elpe + 1);
```

Comment:
What if the size exceeded the pre-allocated buffer size?

Reply:
I see, please raise a error message when it's skipped unexpectedly.

Reply:
done